pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -101,6 +184,35 @@ func newEndpoint(address string, port uint32) (*endpoint.LbEndpoint, error) {
 	return ep, nil
 }
 
+// updateCluster is called from the event (or global cache invalidation) to update
+// the endpoints for the cluster.
+func updateCluster(clusterName string, edsCluster *EdsCluster) {
+	hostname, ports, labels := model.ParseServiceKey(clusterName)
+	instances, err := edsCluster.discovery.mesh.Instances(hostname, ports.GetNames(), labels)
```

Comment:
I would suggest moving the lock above this. If anything, this will populate local client-go cache or other such stuff. It avoids querying the service registry for each caller of updateCluster

Reply:
Well - I considered this, but will make the lock to be held for longer, with unknown effects. 

Added a TODO to evaluate - for now I think I would take the risk of calling multiple times.